js/store.js

```diff
@@ -23,49 +26,89 @@ export const TodoStore = class extends EventTarget {
 				? this.todos.filter((todo) => todo.completed)
 				: this.todos;
 	}
-	_readStorage() {
-		this.todos = JSON.parse(window.localStorage.getItem(this.localStorageKey) || "[]");
+	_handleError = (error) => {
+		this.dispatchEvent(new ErrorEvent('error', {error}));
+	};
+	_readStorage(isNew) {
+		this.todos = [];
+		const loadTodoReferences = async (state) => {
+			await Promise.all(this.todos.map(async (todo, i) => {
+				if (isReference(todo))
+					this.todos[i] = await state.get(todo['@id']);
+			}));
+			console.log(this.todos);
+			this.dispatchEvent(new CustomEvent("save"));
+		};
+		clone(new MemoryLevel, IoRemotes, {
+			'@id': uuid(),
+			'@domain': `${this.id}.todomvc.m-ld.org`,
+			genesis: isNew,
+			io: {uri: `http://${window.location.hostname}:3001`}
+		}).then(async meld => {
+			this.meld = meld;
+			await meld.status.becomes({ outdated: false });
+			meld.read(async state => {
+				this.todos = (await state.get('todos'))?.['@list'] ?? [];
+				await loadTodoReferences(state);
+			}, async (update, state) => {
+				updateSubject({'@id': 'todos', '@list': this.todos}, update);
+				await loadTodoReferences(state);
+			});
+		}).catch(this._handleError);
 	}
```

Comment:
**suggestion:** I don't love how spread out these side effects are—namely, that
in each branch of the `read`, we mutate `this.todo` and then call
`loadTodoReferences()`, which finds that array on its own, mutates its elements,
and then dispatches the `save` event. Consider something more like this, which
does the todo item lookups functionally, and only updates `this.todo` in a
single location, atomically, in the same place where the `save` event is dispatched.

```suggestion
	_readStorage(isNew) {
		this.todos = [];
		const getTodos = (state, todoRefs) =>
			Promise.all(todoRefs.map(async (todoRef) =>
				state.get(todoRef['@id'])
			));
		const setTodos = async (state, todoRefs) => {
			this.todos = await getTodos(state, todoRefs);
			this.dispatchEvent(new CustomEvent("save"));
	  }
		clone(new MemoryLevel, IoRemotes, {
			'@id': uuid(),
			'@domain': `${this.id}.todomvc.m-ld.org`,
			genesis: isNew,
			io: {uri: `http://${window.location.hostname}:3001`}
		}).then(async meld => {
			this.meld = meld;
			await meld.status.becomes({ outdated: false });
			meld.read(async state => {
				await setTodos(state, (await state.get('todos'))?.['@list'] ?? []);
			}, async (update, state) => {
				await setTodos(state, updateSubject({'@id': 'todos', '@list': [...this.todos]}, update)['@list']);
			});
		}).catch(this._handleError);
	}
```

A potential downside to this approach is that the todos won't fill themselves in
as they become available, but we're already not rendering them until they've all
been fetched, so we already have that issue visually, we just have unrendered
data in the model object earlier.

Reply:
As discussed, I'll leave the in-place hydration pattern for now, but make the code more readable and put in a comment about how this will improve.